### PR TITLE
Move Sprite Lab interpreted library to run natively

### PR DESCRIPTION
**Pull request by ajpal**
This moves the sprite lab library to run natively instead of in the interpreter. See https://docs.google.com/document/d/1MLuWzR92baGfjjEd5Whags9z3fw9jGCRCIvFm1kMcYw/edit for the description of the technical approach and https://docs.google.com/document/d/1wcUeKwCHo3dzog_cU1rKuDpKWeG-NCdjycasqFyqtrA/edit for a description of the major changes.
In order to make that possible, there are several important changes:
* Most importantly, the interpreter no longer has access to sprite objects directly. It now interfaces with sprites by either unique identifier or animation name (costume group)
* Custom helper library code and validation code had to be changed for every Spritelab level
* The interpreted library now contains only the draw loop, which is a thin wrapper to the native function `executeDrawLoopAndCallbacks()`, and what I've started calling the "legacy library". These functions are required to keep old blocks working, and we plan to maintain it forever, but we don't plan on making any changes to these functions, so they are staying in the interpreter. Many of them actually just point to the updated versions of the block (ex. `whenDownArrow()` just calls to `keyPressed()`)

Next steps (follow on PRs):
* Unit tests!
* Delete blocks that are not used anywhere and will not work anymore.
* Change generated code for `createNewSprite()` so that it doesn't take `name` as a parameter
* Resolve existing unclear behavior with `when touching` for multi-sprite collisions

### CODE REVIEW

File: apps/src/gamelab/spritelab/Spritelab.js
Comment (jmkulwik) [inline apps/src/gamelab/spritelab/Spritelab.js:2]: Would it make more sense to call this spritelabCommandsTemplate? I'm thinking that if we ever want to completely separate gamelab and spritelab in the future, we would want Spritelab.js to instead include things like 'reset' or 'run'

File: apps/src/gamelab/spritelab/Spritelab.js
Comment (jmkulwik) [inline apps/src/gamelab/spritelab/Spritelab.js]: Looks like name isn't used here.

File: apps/src/gamelab/spritelab/Spritelab.js
Comment (jmkulwik) [inline apps/src/gamelab/spritelab/Spritelab.js]: Is this short for Anonymous? Can we just swap it to 'makeNewSpriteAnonymous' for clarity?

File: apps/src/gamelab/spritelab/Spritelab.js
Comment (jmkulwik) [reply apps/src/gamelab/spritelab/Spritelab.js:2]: Then perhaps make Spritelabe.js just have drawBackground, updateTitle, and reset. And import the commands to a variable.

File: apps/src/gamelab/spritelab/actionCommands.js
Comment (jmkulwik) [inline apps/src/gamelab/spritelab/actionCommands.js]: Would it make sense to check for spriteId and prop here too?

Comment (jmkulwik) [reply apps/src/gamelab/spritelab/actionCommands.js]: (similarly as well for other functions here)

File: apps/src/gamelab/spritelab/actionCommands.js
Comment (jmkulwik) [inline apps/src/gamelab/spritelab/actionCommands.js:70]: haha now I know why we were talking about trig a couple days ago

File: apps/src/gamelab/spritelab/Spritelab.js
Comment (ajpal) [reply apps/src/gamelab/spritelab/Spritelab.js]: Unfortunately, we can't easily change the function name, because that would require changing the block name as well. Changing the block name would break every project that already exists with the `makeNewSpriteAnon` block (which we definitely don't want to do haha)

+1 to your point, though, and definitely something to keep in mind when naming new blocks in the future.

File: apps/src/gamelab/spritelab/actionCommands.js
Comment (jmkulwik) [inline apps/src/gamelab/spritelab/actionCommands.js:82]: Should this return after logging?

File: apps/src/gamelab/spritelab/actionCommands.js
Comment (jmkulwik) [inline apps/src/gamelab/spritelab/actionCommands.js]: Angle could be 0 when there is still distance between the two which would cause this to evaluate to false. i.e. 
`Math.atan2(5 - 5, 50 - 0)`

File: apps/src/gamelab/spritelab/Spritelab.js
Comment (ajpal) [reply apps/src/gamelab/spritelab/Spritelab.js]: Yeah this is another quirk of generated code :/
This block generates:
```
var name;
name = createNewSprite(name, animation, location);
```
I don't know of a way to make it *not* pass name into the function as well.
It's probably possible, but I don't want to block this change on that. I'll log a ticket in our backlog though to clean this up for all assignment blocks.

File: apps/src/gamelab/spritelab/spriteCommands.js
Comment (jmkulwik) [inline apps/src/gamelab/spritelab/spriteCommands.js:65]: Can sprite.baseScale ever be 0? Do we need to handle a case where baseScale may not exist?

File: apps/src/gamelab/spritelab/spriteUtils.js
Comment (jmkulwik) [inline apps/src/gamelab/spritelab/spriteUtils.js:68]: Perhaps rename or add a comment. I was thinking this would return a list of behaviors for an animation. But it returns a count of animations that have behaviors.

Comment (jmkulwik) [reply apps/src/gamelab/spritelab/spriteUtils.js:68]: On that note, it seems like "Costume" would be more clear than "Animation." Do you know why we use "Animation" for spritelab?

File: apps/src/gamelab/spritelab/spriteUtils.js
Comment (jmkulwik) [inline apps/src/gamelab/spritelab/spriteUtils.js]: This seems like it could introduce bugs if people start to rely on it for things besides overlap.

File: apps/src/gamelab/spritelab/spriteUtils.js
Comment (ajpal) [reply apps/src/gamelab/spritelab/spriteUtils.js:68]: "Animation" is the p5 name, and "Costume" is the product name for animations in sprite lab. Chris suggested using "animation" throughout so that if we someday change what we're calling them in spritelab, it won't require code changes here

File: apps/src/gamelab/spritelab/spriteUtils.js
Comment (jmkulwik) [inline apps/src/gamelab/spritelab/spriteUtils.js]: Perhaps rename singleOrGroup to getSpriteArray?

File: apps/src/gamelab/spritelab/spriteCommands.js
Comment (ajpal) [reply apps/src/gamelab/spritelab/spriteCommands.js:65]: baseScale is guaranteed to exist (it's initialized to 1 in `makeSprite()`), and the only time it's reassigned, it's set to `100 / Math.max(100, height, width)`, so also guaranteed to be not 0.

File: apps/src/gamelab/spritelab/spriteUtils.js
Comment (ajpal) [reply apps/src/gamelab/spritelab/spriteUtils.js:68]: Renaming this function would require changing all the level validation code. I'm happy to do it, but I'd rather do it as a follow-on task once this PR is merged. I'll add a comment for now.

File: dashboard/config/libraries/NativeSpriteLab.interpreted.js
Comment (jmkulwik) [inline dashboard/config/libraries/NativeSpriteLab.interpreted.js:5]: If this is legacy code, do we have a plan to remove it?

File: dashboard/config/libraries/NativeSpriteLab.interpreted.js
Comment (jmkulwik) [inline dashboard/config/libraries/NativeSpriteLab.interpreted.js]: Might be worth adding a return here in case anyone adds something else to the end.

File: dashboard/config/libraries/NativeSpriteLab.interpreted.js
Comment (jmkulwik) [inline dashboard/config/libraries/NativeSpriteLab.interpreted.js:38]: Do both randColor and randomColor get used?

File: dashboard/config/libraries/NativeSpriteLab.interpreted.js
Comment (ajpal) [reply dashboard/config/libraries/NativeSpriteLab.interpreted.js:5]: No :(
Unfortunately, I think the plan is that we need to support these blocks forever

File: dashboard/config/libraries/NativeSpriteLab.interpreted.js
Comment (jmkulwik) [inline dashboard/config/libraries/NativeSpriteLab.interpreted.js]: NIT: Add a space :) (not sure why the linter didn't catch that lol)

File: dashboard/config/libraries/NativeSpriteLab.interpreted.js
Comment (ajpal) [reply dashboard/config/libraries/NativeSpriteLab.interpreted.js:38]: Yes, I think so. Or at least, we have no way to be really sure which function is used where, so the safest option is to just keep them both.

File: dashboard/config/libraries/NativeSpriteLab.interpreted.js
Comment (ajpal) [reply dashboard/config/libraries/NativeSpriteLab.interpreted.js]: ha! No linter in the library editor on levelbuilder. Will do :)

File: apps/src/gamelab/spritelab/Spritelab.js
Comment (jmkulwik) [reply apps/src/gamelab/spritelab/Spritelab.js]: No prob. Add a comment so someone else doesn't delete it :)

File: apps/src/gamelab/spritelab/spriteUtils.js
Comment (jmkulwik) [inline apps/src/gamelab/spritelab/spriteUtils.js]: If I call delete on a spriteId, and then call allSpritesWithAnimation, will it throw an error "Cannot read property getAnimationLabel of undefined"?

Comment (jmkulwik) [reply apps/src/gamelab/spritelab/spriteUtils.js]: Oh wait. Nevermind.

File: apps/src/gamelab/spritelab/spriteUtils.js
Comment (jmkulwik) [inline apps/src/gamelab/spritelab/spriteUtils.js]: Might just be my personal preference here: I tend to find big switch blocks with lots of inline logic a bit difficult to read. I usually prefer moving each of the case blocks to a function so the switch block itself is much more compact. i.e.
```
switch (type) {
  case 'whenpress': return onWhenPress(inputEvent, p5Inst);
...
```

File: apps/src/gamelab/spritelab/spriteCommands.js
Comment (jmkulwik) [reply apps/src/gamelab/spritelab/spriteCommands.js:65]: 👍

File: apps/src/gamelab/spritelab/spriteUtils.js
Comment (jmkulwik) [reply apps/src/gamelab/spritelab/spriteUtils.js:68]: Sounds fine to me. I'm ok with leaving it as-is with a comment.

File: dashboard/config/libraries/NativeSpriteLab.interpreted.js
Comment (jmkulwik) [reply dashboard/config/libraries/NativeSpriteLab.interpreted.js:38]: haha 🤦‍♂ sounds good

File: apps/src/gamelab/spritelab/Spritelab.js
Comment (jmkulwik) [reply apps/src/gamelab/spritelab/Spritelab.js:2]: 👏

File: apps/src/gamelab/spritelab/Spritelab.js
Comment (jmkulwik) [reply apps/src/gamelab/spritelab/Spritelab.js]: 👍

File: apps/src/gamelab/spritelab/Spritelab.js
Comment (jmkulwik) [reply apps/src/gamelab/spritelab/Spritelab.js]: 👍

File: apps/src/gamelab/spritelab/actionCommands.js
Comment (jmkulwik) [inline apps/src/gamelab/spritelab/actionCommands.js]: Can you give 'n' a more descriptive name?

File: apps/src/gamelab/spritelab/actionCommands.js
Comment (jmkulwik) [reply apps/src/gamelab/spritelab/actionCommands.js]: 👍

File: apps/src/gamelab/spritelab/spriteUtils.js
Comment (jmkulwik) [reply apps/src/gamelab/spritelab/spriteUtils.js]: 🎉

File: apps/src/gamelab/spritelab/spriteUtils.js
Comment (ajpal) [reply apps/src/gamelab/spritelab/spriteUtils.js]: Yeah, I'm not really sure how best to avoid that though. renamed to `firedOnceForCollision` so that at least it's not as surprising that it gets reset when sprites stop overlapping?

File: dashboard/config/libraries/NativeSpriteLab.interpreted.js
Comment (jmkulwik) [reply dashboard/config/libraries/NativeSpriteLab.interpreted.js:5]: Meh. Ok. Thanks for the super clear comment so people don't add more.

File: apps/src/gamelab/spritelab/spriteUtils.js
Comment (jmkulwik) [reply apps/src/gamelab/spritelab/spriteUtils.js]: Ahhhhh I get it now. I thought the `if(!overlap)` section was a no-op. But it needs to be able to collide multiple times and reset when it's no longer overlapping. Totally makes sense. Definitely the right way to do this. (just add a quick comment)

File: apps/src/gamelab/spritelab/spriteUtils.js
Comment (jmkulwik) [reply apps/src/gamelab/spritelab/spriteUtils.js]: 💯

**Review by jmkulwik — APPROVED** [review]
🎉 Looks awesome! Make sure you change the PR name so it doesn't say "Do Not Submit" anymore. Also add a description. Other than that, ✈️

**Pull request merged**
